src/components/ResizeDetector.js

```diff
@@ -0,0 +1,204 @@
+    this.resizeHandler = handle
+      ? handle(this.createResizeHandler, refreshRate, refreshOptions)
+      : this.createResizeHandler;
+
+    this.resizeObserver = new ResizeObserver(this.resizeHandler);
+  }
+
+  componentDidMount() {
+    this.toggleObserver(true);
+  }
+
+  componentWillUnmount() {
+    this.toggleObserver(false);
+
+    this.rafClean();
+
+    this.cancelHandler();
+
+    this.unmounted = true;
+  }
+
+  cancelHandler = () => {
+    const shouldCancel = !!(this.resizeHandler && this.resizeHandler.cancel);
+    if (shouldCancel) {
+      // cancel debounced handler
+      this.resizeHandler.cancel();
+      this.resizeHandler = null;
+    }
+  }
+
+  rafClean = () => {
+    const shouldClean = !!(this.raf && this.raf.cancel);
+    if (shouldClean) {
+      this.raf.cancel();
+      this.raf = null;
+    }
+  }
+
+  toggleObserver = (isOn) => {
+    const element = this.getElement();
+    if (!element) return false;
+
+    const type = isOn ? 'observe' : 'unobserve';
+    const handle = this.resizeObserver[type];
+    if (!handle) return false;
+
+    this.resizeObserver[type](element);
+    return true;
+  }
+
+  getElement = () => {
+    const { querySelector } = this.props;
+
+    const selectedElement = querySelector && document.querySelector(querySelector);
+
+    // eslint-disable-next-line react/no-find-dom-node
+    const currentElement = findDOMNode(this.element);
+    const parentElement = currentElement && currentElement.parentElement;
+
+    const element = selectedElement || parentElement;
+    return element;
+  };
+
+  createUpdater = () => {
+    this.rafClean();
+
+    this.raf = rafSchd(({ width, height }) => {
+      const { onResize } = this.props;
+
+      if (isFunction(onResize)) {
+        onResize(width, height);
+      }
+      if (this.unmounted) {
+        this.setState({ width, height });
+      }
+    });
+
+    return this.raf;
+  }
+
+  createResizeHandler = (entries) => {
+    const {
+      width: widthCurrent,
+      height: heightCurrent,
+    } = this.state;
+    const {
+      handleWidth,
+      handleHeight,
+    } = this.props;
+
+    const updater = this.createUpdater();
+
+    entries.forEach((entry) => {
+      const { width, height } = (entry && entry.contentRect) || {};
+
+      const isWidthChanged = handleWidth && widthCurrent !== width;
+      const isHeightChanged = handleHeight && heightCurrent !== height;
+      const isSizeChanged = isWidthChanged || isHeightChanged;
+
+      const isHaveWindow = checkIsHaveWindow();
+
+      const shouldSetSize = !this.skipOnMount && isSizeChanged && isHaveWindow;
+      if (shouldSetSize) {
+        updater({ width, height });
+      }
+
+      this.skipOnMount = false;
+    });
+  };
+
+  onRef = (el) => {
+    this.element = el;
+  };
+
+  render() {
+    const { width, height } = this.state;
+    const { render, children } = this.props;
+
+    const childProps = { width, height };
+    const isRenderProps = isFunction(render);
```

Comment:
I suggest extracting this logic to separate methods
```javascript
getRenderType = () => {
  const { render, children } = this.props;
  if (isFunction(render))  {
    return 'renderProp';
  }

  if (isFunction(children)) {
    return 'childFunction';
  }

  if (isValidElement(children)) {
    return 'child';
  }

  return 'parent';
}

getTargetComponent = () => {
  const renderType = this.getRenderType();
  
  switch (renderType) {
    case 'renderProp':
      return cloneElement(render(childProps), { key: 'resize-detector' });
    case '':
      return ...
    case '':
      return ...
  }
}

render() {
  return <Reference ref={this.onRef}>{this.getTargetComponent()}</Reference>
}
```

IMO it's better when you have single responsibility methods that describe some atomic piece of logic.

Reply:
I updated it